[DAGCombiner] Add generic DAG combine for ISD::PARTIAL_REDUCE_MLA

JamesChesterman opened the pull request:
Add generic DAG combine for ISD::PARTIAL_REDUCE_U/SMLA nodes. Transforms the DAG from:
PARTIAL_REDUCE_MLA(Acc, MUL(EXT(MulOpLHS), EXT(MulOpRHS)), Splat(1)) to
PARTIAL_REDUCE_MLA(Acc, MulOpLHS, MulOpRHS).

JamesChesterman committed:
[DAGCombiner] Add generic DAG combine for ISD::PARTIAL_REDUCE_MLA

Add generic DAG combine for ISD::PARTIAL_REDUCE_U/SMLA nodes.
Transforms the DAG from:
PARTIAL_REDUCE_MLA(Acc, MUL(EXT(MulOpLHS), EXT(MulOpRHS)), Splat(1))
to
PARTIAL_REDUCE_MLA(Acc, MulOpLHS, MulOpRHS).

JamesChesterman committed:
Split the DAG combine into two.

Also make sure the DAG combine is only done when the action for
partial reductions have a type combination which is either Legal
or Custom.
This ensures that the combines are not performed only for the
resulting DAG to be expanded, as this leads to worse Code Gen.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,50 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
 
+SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
+  // Makes PARTIAL_REDUCE_MLA(Acc, MUL(EXT(MulOpLHS), EXT(MulOpRHS)), Splat(1))
+  // into PARTIAL_REDUCE_MLA(Acc, MulOpLHS, MulOpRHS)
```
Done.
Added a condition that checks whether the type combination for the Partial Reduction operation is Legal or Custom. The DAG combines will only be done if this is true.
This is to prevent the DAG combines from happening just for the resulting DAG to be expanded, as this leads to worse code gen. 
For now, the combines will not be performed This is because setting the partial reduction action to Legal / Custom for certain types will make instruction selection happen, which is not implemented yet, so an error occurs.

JamesChesterman replied:
Nevermind sorry I've spotted a problem with what I've done.
At the start of `visitPARTIAL_REDUCE_MLA` I've checked whether `N->getOperand(1)` is legal or custom, when really I should be checking the unextended input type to the `MUL` operation, as this is what really matters when lowering.
The only issue is, to check this, I would need to make sure that there are extends on both of the MUL operands and that these are the same signedness etc before performing the `MUL` fold, which is repeating code a bit / I'd move a lot of code from the `foldExtendPARTIAL_REDUCE_MLA` function to before the `foldMulPARTIAL_REDUCE_MLA` function, which might defeat the purpose of it.
I would maybe prefer to just revert back to having the one function for the whole combine, as the `MUL` fold should not be done on its own now, would you be okay with this?

MacDue replied:
I'm slightly inclined towards keeping the combine as one function right now. Since, when it's folding the extends and the `mul` at the same time, then there's no issue since the extends the partial reduce expansion will introduce already existed in the input. Folding the `mul` on its own is implicitly adding extends that did not exist previously (if the `mul` you fold into the partial reduce does not have the same element type as the partial reduce accumulator, which is likely).

sdesmalen-arm replied:
Thanks for trying @JamesChesterman. Moving the `mul` into the `partial_reduce_mla` should only be done if the extends come for free (because there is an instruction that does this), or if this doesn't lose the nsw/nuw flags on the mul (otherwise we lose the information to decide that the `mul` can be performed on the smaller type, i.e.  do the extend on the result of the mul, rather than extending the operands to the accumulator/result type). With that in mind, I agree it is probably easier to do this in one function.

JamesChesterman committed:
Make DAG combine one function again

This is so the MUL fold does not happen unless the extend fold can
be performed.
As otherwise a lot of code would need to be repeated to check that
it can happen.

MacDue commented on llvm/lib/CodeGen/TargetLoweringBase.cpp:
```diff
@@ -836,9 +836,8 @@ void TargetLoweringBase::initActions() {
-    // PartialReduceMLA operations default to expand.
-    setOperationAction({ISD::PARTIAL_REDUCE_UMLA, ISD::PARTIAL_REDUCE_SMLA}, VT,
-                       Expand);
+    for (MVT InputVT : MVT::all_valuetypes())
+      setPartialReduceMLAAction(VT, InputVT, Expand);
```
Is this still needed? This is preventing the new DAG combine from applying, which means it's untested in this PR.

JamesChesterman replied:
If we changed the operation actions to Legal or Custom then the combined node would attempt to be lowered, which is not implemented yet. I can tell it works because I have a patch that will be applied after this in which it is successfully lowered to a dot product instruction (because of the successful DAG combine). Unless there would maybe be some other way to test it?

MacDue replied:
Does the DAG combine need to depend on things being legal or custom now though? I thought that was only to fix the issue with splitting the combine into two functions?

MacDue replied:
This bit:
```
  // Only perform the DAG combine if there is custom lowering provided by the
  // target
  if (!TLI.isPartialReduceMLALegalOrCustom(N->getValueType(0), MulOpLHSVT))
    return SDValue();
```

JamesChesterman replied:
When the combine is performed and then the combine function is performed, the code generation for the NEON cases is quite a bit longer. There is even a test case that gets longer when the flag `-aarch64-enable-partial-reduce-nodes` is not present.

MacDue replied:
Are you planning on implementing the node lowering before landing this PR?

JamesChesterman replied:
I've implemented the node lowering for the dot products and it works fine. But I was going to merge it after this PR as the combine needs to happen for the lowering to happen properly.

MacDue replied:
Is there an order of PRs which means each PR can be tested in isolation? If not, I think it'd be best to have the changes all in one PR, so we don't merge code that does not have tests to defend it.

JamesChesterman replied:
OK, I could move the dot product lowering changes into this PR I think. Then legalisation, then lowering for USDOTs could come later as these would all change tests as a result of their implementations.

MacDue replied:
Discussed this a little offline. I think it's okay to remove the  `isPartialReduceMLALegalOrCustom` check, and accept a slight regression (so we can test the DAG combine), given we know we're going to land a fix shortly after with the node lowerings.

JamesChesterman replied:
Done

JamesChesterman committed:
Remove isLegalOrCustom check from DAG combine

This makes it so the changes are reflected in the tests, so that we can tell the DAG combine is actually happening.
It has been replaced with a FIXME note saying to potentially add it back in when the rest of the implementation is complete.

MacDue commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -1639,6 +1639,25 @@ class TargetLoweringBase {
            getCondCodeAction(CC, VT) == Custom;
   }
 
+  /// Return how a PARTIAL_REDUCE_U/SMLA node with Acc type AccVT and Input type
+  /// InputVT should be treated. Either it's legal, needs to be promoted to a
+  /// larger size, needs to be expanded to some other code sequence, or the
+  /// target has a custom expander for it.
+  LegalizeAction getPartialReduceMLAAction(EVT AccVT, EVT InputVT) const {
+    unsigned AccI = (unsigned)AccVT.getSimpleVT().SimpleTy;
+    unsigned InputI = (unsigned)InputVT.getSimpleVT().SimpleTy;
+    assert(AccI < MVT::VALUETYPE_SIZE && InputI < MVT::VALUETYPE_SIZE &&
+           "Table isn't big enough!");
+    return PartialReduceMLAActions[AccI][InputI];
+  }
+
+  /// Return true if a PARTIAL_REDUCE_U/SMLA node with the specified types is
+  /// legal or custom for this target.
+  bool isPartialReduceMLALegalOrCustom(EVT AccVT, EVT InputVT) const {
+    return getPartialReduceMLAAction(AccVT, InputVT) == Legal ||
```
Should these changes (and setPartialReduceMLAAction) be moved to the PR that adds the lowering, as they're not really used in this PR?

JamesChesterman replied:
Done

JamesChesterman committed:
Remove unnecessary functions for getting PR operation action

MacDue reviewed (approved):
LGTM, thanks :+1: I think the regressions should be fixed after the lowering changes.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,54 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+  if (Op1->getOpcode() != ISD::MUL)
+    return SDValue();
+
+  APInt ConstantOne;
+  if (!ISD::isConstantSplatVector(N->getOperand(2).getNode(), ConstantOne) ||
+      !ConstantOne.isOne())
+    return SDValue();
```
This is not the same, `AllOnes` in this case means `~0`, not `1`.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,54 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+      !ConstantOne.isOne())
+    return SDValue();
+
+  SDValue ExtMulOpLHS = Op1->getOperand(0);
+  SDValue ExtMulOpRHS = Op1->getOperand(1);
```
nit: could you give this the shorter names `LHS` and `RHS`?

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,54 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+      !ISD::isExtOpcode(ExtMulOpRHSOpcode))
+    return SDValue();
+
+  SDValue MulOpLHS = ExtMulOpLHS->getOperand(0);
+  SDValue MulOpRHS = ExtMulOpRHS->getOperand(0);
```
nit: and `LHSExtOp` and `RHSExtOp` would probably be more appropriate.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,54 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+  if (MulOpLHSVT != MulOpRHS.getValueType())
+    return SDValue();
+
+  // FIXME: Add a check to only perform the DAG combine if there is lowering
+  // provided by the target
```
Is it worth adding a TLI hook already in this patch that takes the operands and only returns true if all types involved are all legal? (this can then be extended in a later patch to handle some more cases, anticipating widening and promotion) Or do you think this is better done in a follow-up patch?

JamesChesterman committed:
Rename variables and move a comment.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,54 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+
+  bool LHSIsSigned = ExtMulOpLHSOpcode == ISD::SIGN_EXTEND;
+  bool RHSIsSigned = ExtMulOpRHSOpcode == ISD::SIGN_EXTEND;
+  if (LHSIsSigned != RHSIsSigned)
```
In a future patch, I think that the TLI hook will prevent type combinations that are inappropriate for the combine from being changed in this function.

MacDue replied:
I think DAG combine should be correct regardless of the legalization actions. I think a simple check like:
```c++
bool NodeIsSigned = N->getOpcode() == ISD::PARTIAL_REDUCE_SMLA;
if (LHSIsSigned != NodeIsSigned 
  && (LHS.getValueType() != Acc.getValueType() || RHS.getValueType() != Acc.getValueType()) {
  return SDValue();
}
```
Would do :)

JamesChesterman replied:
Done, sorry about that, I misunderstood.

JamesChesterman committed:
Add an additional check into the DAG combine

JamesChesterman committed:
Rename variables to match those in the function

MacDue reviewed (approved):
This still LGTM, but wait for @sdesmalen-arm

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,61 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+  SDValue Op1 = N->getOperand(1);
+  SDValue Op2 = N->getOperand(2);
+
+  if (Op1->getOpcode() != ISD::MUL)
+    return SDValue();
+
+  APInt ConstantOne;
+  if (!ISD::isConstantSplatVector(Op2.getNode(), ConstantOne) ||
+      !ConstantOne.isOne())
+    return SDValue();
```
nit:
```suggestion
APInt ConstantOne;
if (Op1->getOpcode() != ISD::MUL ||
    !ISD::isConstantSplatVector(Op2.getNode(), ConstantOne) ||
    !ConstantOne.isOne())
  return SDValue();
```

JamesChesterman replied:
Done

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,61 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+  SDValue LHSExtOp = LHS->getOperand(0);
+  SDValue RHSExtOp = RHS->getOperand(0);
+  EVT LHSExtOpVT = LHSExtOp.getValueType();
+  if (LHSExtOpVT != RHSExtOp.getValueType())
```
nit: at this point you can fold in the `LHSIsSigned != RHSIsSigned` check into this condition, by just checking if the opcodes match.
```suggestion
  if (LHSExtOpVT != RHSExtOp.getValueType() || LHSOpcode != RHSOpcode)
```

JamesChesterman replied:
Done

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,61 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+  if (LHSIsSigned != RHSIsSigned)
+    return SDValue();
+
+  bool NodeIsSigned = N->getOpcode() == ISD::PARTIAL_REDUCE_SMLA;
```
Could you add a brief comment explaining that for a 2-stage extend the signedness of the extends must match.

JamesChesterman replied:
Done (added before the new check for `LHSOpcode != RHSOpcode`)

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,61 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+
+  bool NodeIsSigned = N->getOpcode() == ISD::PARTIAL_REDUCE_SMLA;
+  EVT AccElemVT = Acc.getValueType().getVectorElementType();
+  if (LHSIsSigned != NodeIsSigned &&
+      (Op1.getValueType().getVectorElementType() != AccElemVT ||
+       Op2.getValueType().getVectorElementType() != AccElemVT))
```
You only need to check this for `Op1`, because the types of `Op1` and `Op2` must match.
```suggestion
  if (LHSIsSigned != NodeIsSigned &&
      Op1.getValueType().getVectorElementType() != AccElemVT)
```

JamesChesterman replied:
Done

sdesmalen-arm reviewed (commented):
Few more small comments, other than that it looks fine.

JamesChesterman committed:
Simplify DAG combine

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12497,6 +12501,58 @@ SDValue DAGCombiner::visitMHISTOGRAM(SDNode *N) {
+  if (!ISD::isExtOpcode(LHSOpcode) || !ISD::isExtOpcode(RHSOpcode))
+    return SDValue();
+
+  // For a 2-stage extend the signedness of both of the extends must be the
+  // same. This is so the node can be folded into only a signed or unsigned
+  // node.
```
You've added the comment to the wrong condition (it must be moved to the one below).

JamesChesterman replied:
Done

JamesChesterman committed:
Move comment

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll:
```diff
@@ -258,19 +274,27 @@ define <4 x i64> @sdot_8to64(<4 x i64> %acc, <16 x i8> %a, <16 x i8> %b){
-; CHECK-NODOT-NEXT:    smull2 v2.8h, v2.16b, v3.16b
-; CHECK-NODOT-NEXT:    sshll v3.4s, v4.4h, #0
-; CHECK-NODOT-NEXT:    sshll v5.4s, v2.4h, #0
+; CHECK-NODOT-NEXT:    sshll v4.8h, v3.8b, #0
```
I know these are regressions, but they'll be addressed by follow-up patches that further improve this code-gen.

kazutakahirata commented:
@JamesChesterman I've reverted this PR due to a failure in `check-llvm`.  I am happy to try your revised patch.  Thanks!

JamesChesterman commented:
Relanded this PR successfully: [https://github.com/llvm/llvm-project/commit/e3c8e17b073bf38e900014ed47230882c407dab8](url)

kazutakahirata commented:
@JamesChesterman Thank you for fixing this!
